lambdaclass/mirra_backend
# [GH-796] Quests missing features

## Motivation

We need to rework how we communicate and complete daily quests, also add weekly quests:

- We need to send the state and information of the quest that a user has in an endpoint, we'll the user show endpoint for this purpose
- implement *weekly quests* they are the same as daily but you have a whole week to complete them instead of just one day
- Currently, we're completing the daily quest when a match is finished, this is wrong. The daily quests should be claimed through a button in the client. 
- We need an endpoint to be able to generate quest with a request


Closes #796

## Summary of changes
The main objective of this pr is to be able to fully use the quest feature without accessing the elixir terminal, and also implement the weekly quests feature


- Removed quest completion on match registration and added endpoint to complete/claim quests
- Added endpoint to generate user quests for a user
- Added quests to the user when retrieving him through the show endpoint
- Added weekly quest type

## How to test it?

- Enter the game as a guest in Unity and copy the user_id:
<img width="664" alt="Captura de pantalla 2024-07-24 a la(s) 5 10 15 p  m" src="https://github.com/user-attachments/assets/bef03a34-20ce-4220-8227-5b1f11ac6d03">

- Run get through the terminal to see that the user doesn't own any quest
`curl -X GET  http://localhost:4001/curse/users/user_id`
- Give daily quests to the user
`curl -X POST -d type=daily -d amount=3 http://localhost:4001/curse/users/user_id/quest/add_quests`
- Give weekly quests to the user
`curl -X POST -d type=weekly -d amount=3 http://localhost:4001/curse/users/user_id/quest/add_quests`
- Get to see if he has any quest
`curl -X GET  http://localhost:4001/curse/users/user_id`
- Play a match in unity and lose 
- if you run the get endpoint again you should see that the quest has some progress depending on your actions in the match, there's a quest that you need to lose a match to complete, copy that quest "id" field and run the following command:
`curl -X GET  http://localhost:4001/curse/users/user_id/quest/quest_id/complete_quest`
You should get a `quest completed` message the first time and an error message the second time. The same should happen to the rest of the unfinished quests


## Checklist
